### DISCUSSION TITLE
[RFC] Vouchers unlocking promotions

### DISCUSSION THREAD

**Discussion by IKarbowiak** [saleor/saleor#14409] [RFC]
## Problem

We want to have an option to create multiple codes to one voucher configuration and to introduce the option to apply vouchers codes only when specific conditions are met defined by the promotion rules.

---

## 1. General Assumptions

- Introduce new `VoucherCode` model that will keep voucher codes
- Allow defining rules only with `checkoutAndOrderPredicate` conditions for vouchers
- The rules conditions will be checked when the voucher code will be added to checkout
- Instead of `VoucherChannelListing` rules will be used to define the channel availability (deprecate channel listings mutations and types)
- The voucher will need to have rules with channels defined to be available on specific channels.
- Allow exporting voucher codes to CSV
- Allow bulk voucher codes creation
- Add webhooks for voucher codes changes
- Add events for voucher codes changes

### Acceptance Criteria

- [ ] Adding a rule to voucher that specifies `cataloguePredicate` will raise an error
- [ ] The voucher is applied in checkout only when the conditions from rules are fulfilled
- [ ] The voucher from discount with voucher codes without any predicates should be always applied on checkout from matching channels

### Mutations
Update voucher mutations:

**VoucherInput**:
- deprecate `code` field
- add `codes: List[str]` field
- deprecate `type` and `discountValueType` fields (the discount type will be defined in the rules)
- deprecate `products`, `variants`, `collections`, `categories` fields (will be defined in rules)

```graphql
input VoucherInput @doc(category: "Discounts") {
    code: String @deprecate
    products: [ID!] @deprecate
    variants: [ID!] @deprecate
    collections: [ID!] @deprecate
    categories: [ID!] @deprecate
 
    # new fields
    codes: [String!]
    rules: [VoucherRuleInput!]
   
}

input VoucherRuleInput @doc(category: "Discounts") {
  name: String
  description: JSON
  checkoutAndOrderPredicate: CheckoutAndOrderPredicateInput
  rewardValueType: RewardValueTypeEnum
  rewardValue: PositiveDecimal
  channels: [ID!]
}

```

Rename promotion rule mutations:
- `promotionRuleCreate` --> `discountRuleCreate`
- `promotionRuleUpdate` --> `discountRuleUpdate`
- `promotionRuleDelete` --> `discountRuleDelete`

Rename `PromotionRuleCreateInput` to `DiscountRuleCreateInput`
- add `voucher: ID` field

**DiscountRuleCreate** mutation:
- require `voucher` or `promotion` id
- if the `voucher` and `cataloguePredicate` is provided - raise a `ValidationError` (only `checkoutAndOrder` predicate can be provided for vouchers)

**DiscountRuleUpdate** mutation:
- if the rule is voucher rule - raise a `ValidationError` when `cataloguePredicate` is provided

#### Export
Add mutations for exporting voucher codes

#### Deprecate:
- `VoucherChannelListing` type
- fields on `Voucher` type:
  - `code`
  - `used`
  - `products`, `variants`, `collections`, `categories`
  - `type`
  - `discount_value`
  - `discount_value_type`
  - `countries` ❓ is it enough to define countries by channels from rules ❓ 
  - `channel_listings`
  
 - For `code`, `used` we should take first voucher code and return those valued for this instance.
 - For rules related fields, we should get the first rules for given channel and return corresponding fields.
 - Add above information in field descriptions.

#### Rules changes 
Rename `PromotionRule` to `DiscountRule`

#### Voucher changes
- Add `VoucherCode` type
- Extend `Voucher` with `codes`, `rules` fields

```graphql
type VoucherCode @doc(category: "Discounts") {
  code: String!
  used: Int!
}

type Voucher {
   codes: [VoucherCode!]
   rules: [DiscountRule!]
}

```

#### Voucher changes
Create `VoucherCode` model with relation to voucher
```python
class VoucherCode:
    code = models.CharField(max_length=255, unique=True, db_index=True)
    used = models.PositiveIntegerField(default=0, editable=False)
    voucher = models.ForeignKey(Voucher)
```

**Voucher** changes:
- drop `code` from `Voucher` model
- drop `used` from `Voucher` model
- drop `type` from `Voucher` model 
- drop `products`, `variants`, `collections`, `categories` relations
- drop `countries` fields ❓ is it enough to define countries by channels from rules ❓ 
- drop `VoucherChannelListing`

#### Rules:
- Allow assigning voucher to rules.
- Rename `PromotionRule` to `DiscountRule`
```python
class DiscountRule:
    voucher = models.ForeignKey(
        Voucher, on_delete=models.CASCADE, related_name="rules"
    )
```

#### Migrations:
- migrate current voucher codes and `used` field to `VoucherCode` objects
- migrate voucher channel listings, voucher type, and  `products`, `variants`, `collections`, `categories` relations to promotion rules (as it's done for sales)

## 4. Calculations
- When the `checkoutAddPromoCode` with voucher code is called, we need to check if the given checkout fits the voucher conditions (if `checkoutAndOrder` predicates are provided). When the conditions are not fulfilled - raise a `ValidationError`, otherwise apply the discount.

## 5. To Do list

```[tasklist]
### Tasks
- [ ] Update voucher resolvers to match the new flow
- [ ] Update checkout calculations to apply the voucher only when the rules conditions are met
- [ ] Add zero downtime support
```

## 6. To Test

- [ ]
- [ ]

**Comment by korycins** [comment]
@IKarbowiak What will return deprecated fields on Voucher like `code` in case of having multiple codes? Will it return the first code? Or null (if it is a nullable field)? This should be mentioned in the description after deprecating.
Same questions to the rest of the fields, like `type` or `used`

**Comment by IKarbowiak** [comment]
Good question here. I think yes, the `code` will return first code, same `used` will returned `used` from the first `code`. What about type, we will take this from the first `rule`. I will update the description with those assumptions. Thanks @korycins
